### PR TITLE
Avoid using StandardCharset in Java code

### PR DESCRIPTION
**Pull request by minborg**
According to a comment in `java.nio.charset.StandardCharsets`, the use of this class should be avoided from elsewhere in the `jdk.base` module. This PR proposes removing all such direct use of said class in Java code (but not in JavaDocs).

```
public final class StandardCharsets {

    // To avoid accidental eager initialization of often unused Charsets
    // from happening while the VM is booting up, which may delay
    // initialization of VM components, we should generally avoid depending
    // on this class from elsewhere in java.base.

```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign.git pull/864/head:pull/864` \
`$ git checkout pull/864`

Update a local copy of the PR: \
`$ git checkout pull/864` \
`$ git pull https://git.openjdk.org/panama-foreign.git pull/864/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 864`

View PR using the GUI difftool: \
`$ git pr show -t 864`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/864.diff">https://git.openjdk.org/panama-foreign/pull/864.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/panama-foreign/pull/864#issuecomment-1677177445)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/StringSupport.java [hunk 1]
```diff
@@ -241,15 +241,14 @@ public int terminatorCharSize() {
...
-                throw new UnsupportedOperationException("Unsupported charset: " + charset);
-            }
+            // Switching on the charset names rather than specific instances of
+            // `Charset` avoids loading the class `StandardCharsets`
+            return switch (charset.name()) {
+                case "UTF-8", "ISO8859_1", "US-ASCII" -> CharsetKind.SINGLE_BYTE;
+                case "UTF-16LE", "UTF-16BE", "UTF-16" -> CharsetKind.DOUBLE_BYTE;
+                case "UTF-32LE", "UTF-32BE", "UTF-32" -> CharsetKind.QUAD_BYTE;
+                default -> throw new UnsupportedOperationException("Unsupported charset: " + charset);
+            };
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/StringSupport.java:251]: You might want to run the StrLenTest benchmark to see if this change has any effect on it.

**Comment by minborg** [comment]
I've ran some benchmarks:

```
Benchmark                       Mode  Cnt  Score   Error  Units
SwitchBench.ifOnInstancesUtf32  avgt   30  0.654 ? 0.038  ns/op
SwitchBench.ifOnInstancesUtf8   avgt   30  0.576 ? 0.018  ns/op
SwitchBench.ifOnStringsUtf32    avgt   30  3.923 ? 0.015  ns/op
SwitchBench.ifOnStringsUtf8     avgt   30  0.597 ? 0.018  ns/op
SwitchBench.optimizedUtf32  avgt   30  0.618 ? 0.040  ns/op
SwitchBench.optimizedUtf8   avgt   30  0.576 ? 0.016  ns/op
SwitchBench.switchOnNamesUtf32  avgt   30  1.059 ? 0.005  ns/op
SwitchBench.switchOnNamesUtf8   avgt   30  1.058 ? 0.004  ns/op
```

And the string switch is a bit slow so I will change to another "optimized" solution.

**Review by mcimadamore — APPROVED** [review]
Looks good

**Pull request closed without merging**